docs(langsmith): add Author assertions section to annotation queues (#3923)
langchain-ai/docs: src/langsmith/assertions.mdx

---
title: Use assertions
sidebarTitle: Use assertions
description: Capture free-form acceptance criteria as dataset examples by writing assertions while reviewing runs in an annotation queue.
icon: "checklist"
---

> Assertions turn a reviewer's English-language standards into an automated check. They are short, free-form claims about what a correct answer should or shouldn't do. You write them while reviewing a run in a [single-run annotation queue](/langsmith/annotation-queues#single-run-annotation-queues), and LangSmith saves each one on a [dataset example](/langsmith/example-data-format). Any [offline evaluator](/langsmith/evaluation-concepts#offline-evaluations) can then check whether new outputs from your application satisfy each claim.

Reviewer: should or shouldn't 'include' probably makes more sense

Reviewer 2: updated

Use assertions when:

- The run's actual output is wrong, and you'd rather describe what a correct answer looks like than write one by hand.
- You want to capture acceptance criteria in plain English without leaving the review flow.

<Note>
Assertions are available on [single-run annotation queues](/langsmith/annotation-queues#single-run-annotation-queues). [Pairwise queues](/langsmith/annotation-queues#pairwise-annotation-queues) are unchanged. Assertions are currently UI-only; the [SDK guide](/langsmith/annotation-queues-sdk) covers programmatic access to feedback configs and queue rubrics, not assertions.
</Note>

> ## Add assertions

Reviewer 3: If we were to add a screenshot, I think this section would benefit the most from it.

Reviewer 2: added screenshot to step 5

> 1. In the LangSmith UI, navigate to **Annotation Queues** in the left sidebar. Open a single-run queue and select a run.

Reviewer 3: can we link to the annotation queues tab here? Anywhere where the LangSmith UI is mentioned, it should be a best practice to link to the console or the specific page in the console. If we do this, we might not need to add the screenshot

Reviewer 2: I updated with link to [LangSmith UI](https://smith.langchain.com).
I could not update with a link directly to the specific page in the console, because the url has the workspace id, such as: https://smith.langchain.com/o/ebbaf2eb-769b-4505-aca2-d11de10372a4/annotation-queues. So I will add a screenshot.

1. In the side panel, find the **Assertions** section below **Feedback**.
1. Click **+ Add** to create an assertion row.
1. Enter a **key** that summarizes the claim (for example, `must_cite_source`, `must_not_invent_url`) and a one-sentence **comment** describing the claim.

    The key is free-form. The `must_` / `must_not_` prefixes are just a naming convention; LangSmith doesn't treat them specially.

1. Repeat for each criterion you want to capture.

> The run editor shows the run's inputs and outputs alongside the assertions side panel. As soon as you add at least one assertion, the run editor's **Outputs** panel switches from the run's actual output to a read-only preview of the assertions you've added. This preview is what gets saved to the dataset. The run's actual output is not saved, because assertions describe what a correct answer must look like, not what this run produced.

Reviewer: describe what a correct answer should include*

Reviewer 2: updated

> The run editor shows the run's inputs and outputs alongside the assertions side panel. As soon as you add at least one assertion, the run editor's **Outputs** panel switches from the run's actual output to a read-only preview of the assertions you've added. This preview is what gets saved to the dataset. The run's actual output is not saved, because assertions describe what a correct answer must look like, not what this run produced.

Reviewer 3: ```suggestion
    The run editor shows the run's inputs and outputs alongside the assertions side panel. As soon as you add at least one assertion, the run editor's **Outputs** panel switches from the run's actual output to a read-only preview of the assertions you've added. This preview is what gets saved to the dataset. The run's actual output is not saved, because assertions describe what a correct answer must look like, not what this run produced.
```

Reviewer 2: updated

> You can keep editing the run's **Inputs** at any time, for example to refine the prompt before saving the example. The **Outputs** panel stays locked to the assertion preview while any assertions remain.

Reviewer 3: Should be indented too

Reviewer 2: updated

> ## Save the example

Reviewer 3: Should save the example just be part of the `Add assertions` section? Is there a benefit to having it separate?

Reviewer 2: added as step to Add assertions section

Click **Add to Dataset & Next** in the side panel footer (keyboard shortcut: <kbd>⌘ Enter</kbd> on macOS or <kbd>Ctrl Enter</kbd> elsewhere). LangSmith adds the current run to the queue's [default dataset](/langsmith/annotation-queues#basic-details), or prompts you to pick one if no default is configured. The queue then moves you to the next run.

> The saved example's `outputs` field has this shape (YAML shown for readability; LangSmith stores it as JSON):

Reviewer 3: ```suggestion
The saved example's `outputs` field is stored as JSON,. For example:
```
I don't think the yaml example is needed-- JSON is relatively readable and is how it's actually

Reviewer 2: removed YAML

<CodeGroup>

```json JSON
{
  "assertions": [
    {
      "key": "must_cite_source",
      "comment": "The response cites the source URL it is drawing from."
    },
    {
      "key": "must_not_invent_url",
      "comment": "The response does not include URLs that do not appear in the inputs."
    }
  ]
}
```

> ```yaml YAML

Reviewer 3: I would remove the YAML, idk how helpful it is. Can also remove the codegroup then

Reviewer 2: removed YAML and codegroup
